Source/Core/Core/IOS/Network/KD/NetKDRequest.cpp

```diff
@@ -211,6 +304,82 @@ void NetKDRequestDevice::LogError(ErrorType error_type, s32 error_code)
   m_scheduler_buffer[2] = Common::swap32(new_code);
 }
 
+NWC24::ErrorCode NetKDRequestDevice::KDCheckMail(u32* _mail_flag, u32* _interval)
+{
+  m_scheduler_buffer[4] = Common::swap32(2);
+  u64 random_number{};
+  Common::Random::Generate(&random_number, sizeof(u64));
+  const std::string form_data(
+      fmt::format("mlchkid={}&chlng={}", m_config.GetMlchkid(), random_number));
+  const Common::HttpRequest::Response response = m_http.Post(m_config.GetCheckURL(), form_data);
+
+  if (!response)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "NET_KD_REQ: IOCTL_NWC24_CHECK_MAIL_NOW: Failed to request data at {}.",
+                  m_config.GetCheckURL());
+    return NWC24::WC24_ERR_SERVER;
+  }
+
+  const std::string response_str = {response->begin(), response->end()};
+  const std::string code = GetValueFromCGIResponse(response_str, "cd");
+  if (code != "100")
+  {
+    ERROR_LOG_FMT(
+        IOS_WC24,
+        "NET_KD_REQ: IOCTL_NWC24_CHECK_MAIL_NOW: Mail server returned non-success code: {}", code);
+    return NWC24::WC24_ERR_SERVER;
+  }
+
+  const std::string server_hmac = GetValueFromCGIResponse(response_str, "res");
+  const std::string mail_flag = GetValueFromCGIResponse(response_str, "mail.flag");
+  const std::string interval = GetValueFromCGIResponse(response_str, "interval");
+  DEBUG_LOG_FMT(IOS_WC24, "NET_KD_REQ: IOCTL_NWC24_CHECK_MAIL_NOW: Server HMAC: {}", server_hmac);
+
+  // On a real Wii, a response to a challenge is expected and would be verified by KD.
+  const std::string hmac_message =
+      fmt::format("{}\nw{}\n{}\n{}", random_number, m_config.Id(), mail_flag, interval);
+  std::vector<u8> hashed(20);
+  Common::HMAC::HMACWithSHA1(
+      MAIL_CHECK_KEY,
+      std::span<const u8>(reinterpret_cast<const u8*>(hmac_message.data()), hmac_message.size()),
+      hashed.data());
+
+  if (Common::BytesToHexString(hashed) != server_hmac)
```

Comment:
Is the server guaranteed to return the `server_hmac` in lowercase? If you want to avoid parsing the hex string you probably still should use a case-insensitive comparison -- there should be a function in StringUtil for that.

Reply:
Within KD, strncmp which case sensitive is used to compare both HMAC's. Both RiiConnect24 and WiiLink both return `server_hmac` as lowercase and they both work.

Reply:
Well that's weird enough that I think it's worth leaving a comment about it, but yeah okay then this is fine.